tracer/src/instruction/inline_sha256/mod.rs

```diff
@@ -0,0 +1,516 @@
+        self.add(Reg(self.w(-7)), Reg(self.w(-16)), self.w(-16));
+        // Calculate σ₁(W[t-2])
+        self.sha_word_sigma_1(self.w(-2), ss[0], ss[1]);
+        // Add σ₁ to W[t-16] to get final W[t]
+        self.add(Reg(self.w(-16)), Reg(ss[0]), self.w(-16));
+    }
+
+    /// Computes sha256 Ch function
+    /// Ch(E, F, G) = (E and F) xor ((not E) and G)
+    fn sha_ch(&mut self, rs1: Value, rs2: Value, rs3: Value, rd: usize, ss: usize) -> Value {
+        let e_and_f = self.and(rs1, rs2, ss);
+        let neg_e = self.xor(rs1, Imm(u32::MAX as u64), rd);
+        let neg_e_and_g = self.and(neg_e, rs3, rd);
+        self.xor(e_and_f, neg_e_and_g, rd)
+    }
+
+    /// Computes sha256 Maj function: Maj(A, B, C) = (A and B) xor (A and C) xor (B and C)
+    fn sha_maj(&mut self, rs1: Value, rs2: Value, rs3: Value, rd: usize, ss: usize) -> Value {
+        let b_and_c = self.and(rs2, rs3, ss);
+        let b_xor_c = self.xor(rs2, rs3, rd);
+        let a_and_b_xor_c = self.and(rs1, b_xor_c, rd);
+        self.xor(b_and_c, a_and_b_xor_c, rd)
+    }
+
+    /// Sigma_0 function of SHA256 compression function: Σ₀(x) = ROTR²(x) ⊕ ROTR¹³(x) ⊕ ROTR²²(x)
+    fn sha_sigma_0(&mut self, rs1: Value, rd: usize, ss: usize) -> Value {
+        let rotri_xor = self.rotri_xor_rotri(rs1, 2, 13, rd, ss);
+        let rotri_22 = self.rotri(rs1, 22, ss);
+        self.xor(rotri_xor, rotri_22, rd)
+    }
+
+    /// Sigma_1 function of SHA256 compression function: Σ₁(x) = ROTR⁶(x) ⊕ ROTR¹¹(x) ⊕ ROTR²⁵(x)
+    fn sha_sigma_1(&mut self, rs1: Value, rd: usize, ss: usize) -> Value {
+        let rotri_xor = self.rotri_xor_rotri(rs1, 6, 11, rd, ss);
+        let rotri_25 = self.rotri(rs1, 25, ss);
+        self.xor(rotri_xor, rotri_25, rd)
+    }
+
+    /// sigma_0 for word computation: σ₀(x) = ROTR⁷(x) ⊕ ROTR¹⁸(x) ⊕ SHR³(x)
+    fn sha_word_sigma_0(&mut self, rs1: usize, rd: usize, ss: usize) {
+        self.rotri_xor_rotri(Reg(rs1), 7, 18, rd, ss);
+        let srli = SRLI {
+            address: self.address,
+            operands: FormatI {
+                rd: ss,
+                rs1,
+                imm: 3,
+            },
+            virtual_sequence_remaining: Some(0),
+        };
+        self.sequence.push(srli.into());
+        self.xor(Reg(rd), Reg(ss), rd);
+    }
+
+    /// sigma_1 for word computation: σ₁(x) = ROTR¹⁷(x) ⊕ ROTR¹⁹(x) ⊕ SHR¹⁰(x)
+    fn sha_word_sigma_1(&mut self, rs1: usize, rd: usize, ss: usize) {
+        // We don't need to do Imm shenanigans here since words are always in registers
+        self.rotri_xor_rotri(Reg(rs1), 17, 19, rd, ss);
+        let srli = SRLI {
+            address: self.address,
+            operands: FormatI {
+                rd: ss,
+                rs1,
+                imm: 10,
+            },
+            virtual_sequence_remaining: Some(0),
+        };
+        self.sequence.push(srli.into());
+        self.xor(Reg(rd), Reg(ss), rd);
+    }
+
+    fn lw(&mut self, rs1: usize, imm: i64, rd: usize) {
```

Comment:
to avoid confusion with the actual `imm`
```suggestion
    fn lw(&mut self, rs1: usize, offset: i64, rd: usize) {
```